### ISSUE TITLE
pub: ideally would not depend on barback

### ISSUE THREAD

**Issue by DartBot** [dart-lang/pub#956]
<a href="https://github.com/jmesserly"><img src="https://avatars.githubusercontent.com/u/1081711?v=3" align="left" width="96" height="96"hspace="10"></img></a> **Issue by [jmesserly](https://github.com/jmesserly)**
_Originally opened as dart-lang/sdk#18267_

----

ideally Pub, since it ships in the SDK, would not depend on barback, which is a pkg. Also is generally nice if build tool and pkg manager have a looser API contract, as in the case in other languages.

I know this is crazy hard problem &amp; blocked on lots of other issues, including having an API for dart2js. Perhaps we could file the various other issues, mark this bug as blocked and track progress that way? What do y'all think?

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

If this is possible at all, it's not on the near-term horizon. I don't think it's worth trying to map out all the dependencies necessary for it until we have some time budgeted to start working on them, since things could change in unpredictable ways in the interim.
______
*Removed [Priority-Unassigned](../labels/Priority-Unassigned) label.*
*Added [Priority-Low](../labels/Priority-Low) label.*

**Comment by DartBot** [comment]
<a href="https://github.com/jmesserly"><img src="https://avatars.githubusercontent.com/u/1081711?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [jmesserly](https://github.com/jmesserly)**

----

Yeah, that's fair. I was thinking if there are dependencies, like maybe dart2js, it would be better to notify them so it's on their radar.

One question I have though: what do you mean by "If this is possible at all"? Surely it is possible to have a package manager and build system separate. I can think of many examples from other language ecosystems. Do we have some problem specific to Dart that makes it not possible?

**Comment by DartBot** [comment]
<a href="https://github.com/mezoni"><img src="https://avatars.githubusercontent.com/u/2311549?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [mezoni](https://github.com/mezoni)**

----

>> Surely it is possible to have a package manager and build system separate.

Very reasonable.
As for me, I think that it also should be a separate configuration files.
Not only all-in-one "pubspec.yaml".

This not required a complex "build system".
Only required a configurable "build program". This program should to know how to run "pub".
Or even it should use some public API that written as a "layer" over the "pub" cmd line.

This API should be outside of the Dart SDK but it should use "pub".

Example.

My command line program or your "build program" want install some packages into some directory. How currently this possible? Right.
We should write similar code. Approximately.

&equals;&equals;&equals;&equals;&equals;&equals;&equals;&equals;&equals;
Process.run("$path_to_pub", ["some", "arguments"]).
&equals;&equals;&equals;&equals;&equals;&equals;&equals;&equals;&equals;

But we and me need similar to the following:

import "package:packaging/packaging.dart";

var pm = PackageManager();
var dependencies = [];
dependencies.add(new PackageDependency("foo", version: "1.0.0", source: source));

pm.update(path, dependencies);

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

Calling barback a "build system" isn't really accurate; it does a pretty different set of things than systems like Make or Grunt. Rather than running external tasks that manipulate physical files that are independent of your program, barback runs Dart code on virtual files. More importantly, it often generates the code that is part of your actual program. This means that the act of running a Dart program is often deeply tied to the act of running barback and its transformers. This doesn't necessarily mean that it's impossible to decouple, but it means that it might be impossible.

**Comment by DartBot** [comment]
<a href="https://github.com/jmesserly"><img src="https://avatars.githubusercontent.com/u/1081711?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [jmesserly](https://github.com/jmesserly)**

----

> it often generates the code that is part of your actual program

this can be done in other languages. for example, they run a parser generator that generates C/Java/etc code for inclusion in the program. Another example is reading a UI XML file of some sort to generate a View class that will be part of the program.

I don't see how this ends up being deeply tied to the package manager. Is the issue just that the command happens to be "pub build" instead of "barback build"?

(admittedly I might be stuck in the Polymer/Angular perspective, but we did have Polymer working before and Pub didn't know about it. So that leads me to believe that it is possible. Another data point, polymer-js uses Vulcanizer, which is really simple script built on node.js and grunt: https://github.com/Polymer/vulcanize)

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

It's an issue of transitivity. When all the transformers are local to the entrypoint, it's easy to say "run this script before you run your code." But when you have a transitive dependency five layers deep that uses a transformer to generate one of its source files, it becomes very painful. We expect eventually that many packages will use transformers internally, and it's not really viable to ask users to run "barback build" on their package every time they update their dependencies—to say nothing of the issues of where those built files go, how they're run, etc.

Ultimately, we expect that barback is going to be required to run the majority of real applications, both on the web and on the console. We want users to only have to invoke a single command to run their application, and we want that command to exist in the SDK by default. This means that one way or another, the SDK has to have some sort of dependency on barback.

**Comment by DartBot** [comment]
<a href="https://github.com/jmesserly"><img src="https://avatars.githubusercontent.com/u/1081711?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [jmesserly](https://github.com/jmesserly)**

----

ah, thanks, that's very helpful description of the problem! This discussion is really helpful. At least it's helping me understand. As always apologies for being dense :)

thinking back to various build systems I've worked with, they usually work okay transitively. as long as they can find each "barback.yaml" file (or whatever it is called), there's no problem with also building dependencies. Trying to encapsulate build tasks like that is totally good.

> and it's not really viable to ask users to run "barback build" on their package every time they update their dependencies

ah. That seems easy to fix though. "build" and "serve" are both manual operations. But if we did want it automated, the Editor could be smart about it. For the command like, there could be something like githooks.

> to say nothing of the issues of where those built files go, how they're run, etc.

why is this different from what we do now?

> We want users to only have to invoke a single command to run their application

why can't that not be accomplished with "barback serve"?

> and we want that command to exist in the SDK by default

ah, that's a good point.

But there seems to be various ways to address this. for example, have a "barback" script that simply bootstraps a download, and all of the commands are actually barback-serve, etc. (Git is another great example, where many git commands aren't directly built into the git binary.)

However I think if pub had "bin" packages, it would work too. For example, if I'm developing for node.js I can easily use NPM to install grunt, or insert_favorite_tool_here. Once I install it I have it.

Even if we couldn't find a solution, it wouldn't be very bad. Most users are using the editor, and they get a pubspec for free when they make a new project, and the Editor will even run pub automatically for them. So it would easily pull down barback, and they'd just have it.

Another point to keep in mind, a large number of folks are using Angular and/or Polymer, but we don't ship those in the SDK.

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

> ah. That seems easy to fix though. "build" and "serve" are both manual operations.

They're manual in a sense, but they're also clearly marked as the official way to build and develop your application, respectively. This marking gets muddied if they come from an external package and not from the SDK.

> But if we did want it automated, the Editor could be smart about it.

Bob and I are in the business of designing a pleasant command-line experience. We can't let it be unpleasant and say "we'll fix it in the editor"; in fact, the clear direction we've been getting is to move away from having things that only the editor can do and towards making the editor a thinner shell over command-line functionality.

> For the command like, there could be something like githooks.

At that point, we'd be building another extensibility system that's tightly coupled to pub. It would have all the same coupling problems as barback as we developed it and improved the API.

> why is this different from what we do now?

Now everything is kept in-memory. If building an application is a separate step from running it, the files need to go somewhere.

> why can't that not be accomplished with "barback serve"?

The command needs to work out-of-the-box in the SDK, and it needs to be clear that it's the canonical way to run an application. It's not a good user experience if they start using "pub serve" for their trivial application and then have to switch to "barback serve" when they start using a package that uses a transformer.

> But there seems to be various ways to address this. for example, have a "barback" script that simply bootstraps a download, and all of the commands are actually barback-serve, etc.

That's just another form of dependency from the SDK onto barback.

> Another point to keep in mind, a large number of folks are using Angular and/or Polymer, but we don't ship those in the SDK.

We don't plan to require Angular and Polymer to run literally every Dart application.

Let me summarize our constraints:

\* We expect in the near-to-mid future that many packages, both client- and server-side, will use barback transformers.

\* The way a user interacts with their application shouldn't change when they start using a package that uses barback.

\* The default way to run a Dart application should work with an out-of-the-box SDK.

Given these, I believe we fundamentally need to have a dependency from the SDK onto barback at some level.

It's also worth noting that there is one very straightforward way to stabilize the relationship between the SDK and barback: declare the barback API stable. There is a large but finite amount of engineering effort that will go into doing so, which means that any proposal to expend a large amount of engineering effort reducing the coupling needs to be convincingly less work than just finishing barback.

**Comment by DartBot** [comment]
<a href="https://github.com/jmesserly"><img src="https://avatars.githubusercontent.com/u/1081711?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [jmesserly](https://github.com/jmesserly)**

----

so, it seems like this constraint is the blocker.

> \* The default way to run a Dart application should work with an out-of-the-box SDK.

We need to find a way to loosen it. I offered a few suggestions in my previous message, but I'm open to other ideas.

For what it's worth, the primary framework/components for Dart are not shipped in the SDK. Most of the libraries are not shipped there either. This constraint should not be placed above the more important constraint: a healthy package ecosystem, that doesn't bit rot every 6 weeks.

RE: engineering effort to reduce the coupling. I'll postulate that it's pretty rare to see a successful bit of software that's "finished". There's always a version 2.

Anyway, I agree that finding resources to work on this is an issue.

**Comment by DartBot** [comment]
<a href="https://github.com/blois"><img src="https://avatars.githubusercontent.com/u/2152569?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [blois](https://github.com/blois)**

----

I'm also concerned about tying application execution (pub serve) to the SDK- for scenarios such as an app engine app or a mobile application it seems like it would be ideal to be able to provide a great experience without building it into the SDK.

For example I'd be curious what sort of experience we could build along the lines of Apache's Cordova command-line API without building it into the SDK: https://cordova.apache.org/docs/en/3.1.0/guide_cli_index.md.html

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

> We need to find a way to loosen it. I offered a few suggestions in my previous message, but I'm open to other ideas.

I don't think we do. The barback compatibility policy (https://gist.github.com/nex3/10942218) will mitigate the pain of changes going forward, and requires no additional engineering effort. It matches the compatibility policy of dart:mirrors, which is also under active development, also used by polymer, and more closely tied to the SDK than barback.

> I'm also concerned about tying application execution (pub serve) to the SDK- for scenarios such as an app engine app or a mobile application it seems like it would be ideal to be able to provide a great experience without building it into the SDK.

To be clear, executing an application only requires barback in development. When deployed, "pub build" should be used to remove all external dependencies (except possibly a Dart VM for mobile applications/dartium).

**Comment by DartBot** [comment]
<a href="https://github.com/blois"><img src="https://avatars.githubusercontent.com/u/2152569?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [blois](https://github.com/blois)**

----

> To be clear, executing an application only requires barback in development. When deployed, "pub build" should be used to remove all external dependencies (except possibly a Dart VM for mobile applications/dartium).

The question was more along the lines of "what does the mobile or app engine equivalent of 'pub serve' look like?". Pub serve is a helper for running web applications at development time. How do we scale this to other target runtime environments?

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

That's sort of out-of-scope of this issue, but the short answer is that it will have to use barback somehow, ideally in a way that takes advantage of its incremental building. I believe Kevin and Martin are thinking about AppEngine; I don't know if anyone's thinking about barback and mobile right now.

**Comment by DartBot** [comment]
<a href="https://github.com/jmesserly"><img src="https://avatars.githubusercontent.com/u/1081711?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [jmesserly](https://github.com/jmesserly)**

----

yeah, for now a dart:mirrors policy is sort-of okay. Although it's taking a bad situation -- dart:mirrors breakage -- and doing more of it, which makes it, well doubly bad. :)

but what I'm concerned about is longer term, we're blocked on 1.0 for Angular/Polymer until mirrors and barback are stable.

Even if we declare barback 1.0, it seems really bad if we can never change it. If it was disentangled, we would get total freedom to make changes without fear of breaking things.

That's why we want to keep dart:\* minimal and SDK minimal. So I still maintain this is a very valuable item to solve (once we have time/resources, of course).

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

> but what I'm concerned about is longer term, we're blocked on 1.0 for Angular/Polymer until mirrors and barback are stable.

This is always going to be the case if you use unstable libraries.

> Even if we declare barback 1.0, it seems really bad if we can never change it. If it was disentangled, we would get total freedom to make changes without fear of breaking things.

"Never change it" is inaccurate; we'll be able to safely make additive changes, and we'll be able to make backwards-incompatible changes in Dart 2.0 (although it's an open question how those changes will get tested before the release). Barback, or at least its pub integration, is a core piece of Dart infrastructure, and it makes sense to version it as such.

**Comment by DartBot** [comment]
<a href="https://github.com/nex3"><img src="https://avatars.githubusercontent.com/u/188?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [nex3](https://github.com/nex3)**

----

After hashing this out in person, we've decided to avoid trying to decouple barback from the SDK for now. Instead, we're revising the barback compatibility policy to guarantee that pub's lower bound on barback won't increase until Dart 2.0.
______
*Added [NotPlanned](../labels/NotPlanned) label.*
